proposals/0000-first-class-modules.rst

```diff
@@ -0,0 +1,634 @@
+ambiguity error. [#]_
+
+.. [#] This explication of the semantics uses a local module definition,
+   but we could wire the behaviour in as a special case if we decide
+   not to support local modules.
+
+Module export specifiers
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The previous section allows us to refer to modules by name, but not to
+export those names.
+
+We propose adding a language extension ``ExportModuleNames`` which adds the following
+export specifiers:
+
+- ``module M`` *changes* to exports the name ``M`` as a module binding.
+- ``module M(a, b, .. c)`` exports ``M`` and some of its names.
+- ``module M(..)`` exports ``M`` and all of its names. [#]_
+
+.. [#] See `Do we need additional syntax for re-exporting all bindings from a module?`_ for more discussion.
+
+This ``ExportModuleNames`` behaviour is the “right” behaviour as far as
+this proposal is concerned, but we gate it behind a language extension
+for backwards compatibility.
+
+Local modules
+~~~~~~~~~~~~~
+
+We propose adding a language extension ``LocalModules``. This allows
+module definitions within modules, with the same syntax as in [ghc283]_. Local
+modules are very like file modules, in particular:
+
+- They can have imports.
+
+  - The imports can only target modules which can be referred to
+    with a qualified name in the current scope. Their target module name is *never*
+    interpreted as a dotted name.
+
+- They can have nested module definitions.
+- They can have export specifiers.
+- They can *not* have dotted names.
+
+The rules for local modules are mostly the same as for file modules,
+but:
+
+- A local module definition adds a binding for its declared name to
+  its enclosing scope, as usual for definitions.
+- Bindings from the
+  enclosing scope are in scope in the local module, but are shadowed if
+  there is a local definition of the same name. [#]_
+
+.. [#] See the `Recursion`_ section for some subtleties.
+
+Local modules *cannot* import file modules. This ensures that we can
+always compute the dependency graph of the compilation units of a
+Haskell program by looking at the top-level imports alone. [#]_
+
+.. [#] See `Do we need different syntax for "local" imports?`_ for discussions
+   of the consequences.
+
+Module aliases
+~~~~~~~~~~~~~~
+
+We propose adding a language extension ``ModuleAliases``. This allows a
+new kind of top-level definition: ``module <name> = <qualname>``
+
+A module alias adds a binding in the enclosing scope from the new name
+to the module referred to by the right hand side.
+
+Examples
+--------
+
+Exporting small utility modules
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+::
+
+   module A (f, module Unsafe) where
+
+   f = Unsafe.g …
+
+   module Unsafe where
+       g = … 
+
+Exporting qualified names
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This proposal allows exporting modules, but doesn’t say they can be
+exported under *qualified* names, which is something that people want to
+do. Consider the following `proposed
```

Comment:
> exported under *qualified* names

Are qualified names necessary, or will dotted name suffice?

If dotted names will suffice, I think there's a nice simplification to be made?

A further argument for dotted-instead-of-qualified, is that in the end, the compiler will have to dumb down all this nice structure to something that can be serialised.

And as the implementation experience in https://github.com/ghc-proposals/ghc-proposals/pull/205 has shown, it can be a simple step to:
```
-        tcg_exports :: [AvailInfo],     -- ^ What is exported
+        tcg_exports    :: [AvailInfo],                     -- ^ What is exported
+        tcg_exports_aliases :: [(ModuleName, AvailInfo)], -- ^ ..aliased.
```

Reply:
> A further argument for dotted-instead-of-qualified, is that in the end, the compiler will have to simplify all this nice structure to something that can be serialised efficiently.

Oh I would *really* like GHC to not flatten things with this new design. I think not doing so would basically destroy the proposal. Yes we're making language changes, but I think the associated rich/flexible/structured internal datatype are just as important on their own in the long run.

Reply:
Yes, I don't think we should flatten, and I'm not keen on dotted names. (I thought I had an example of where flattening makes something go wrong, but I can't reconstruct it: I'll think about it.)